[WIP] Add Python language guide (#376)
gitpod-io/website: src/docs/python_in_gitpod.md

## GUI Applications with wxPython

To install wxPython to your repository please add the following to your [`.gitpod.Dockerfile`](https://www.gitpod.io/docs/42_config_docker/). If you don't have one, simply run [`gp init`](https://www.gitpod.io/docs/34_command_line_interface/) and commit the two generated files.

```dockerfile
# This will pull the official Gitpod `vnc` image
# which has much of what you need to start
FROM gitpod/workspace-full-vnc

USER gitpod

# Install wxPython dependencies
RUN sudo apt-get -q update && \
    sudo DEBIAN_FRONTEND=noninteractive apt-get install -yq freeglut3-dev python3.7-dev libpython3.7-dev libgl1-mesa-dev libglu1-mesa-dev libgstreamer-plugins-base1.0-dev libgtk-3-dev libnotify-dev libsdl2-dev libwebkit2gtk-4.0-dev libxtst-dev libgtk2.0-dev && \
    freeglut3-dev \
    python3.7-dev \
    libpython3.7-dev \
    libgl1-mesa-dev \
    libglu1-mesa-dev \
    libgstreamer-plugins-base1.0-dev \
    libgtk-3-dev \
    libnotify-dev \
    libsdl2-dev \
    libwebkit2gtk-4.0-dev \
    libxtst-dev \

>     libgtk2.0-dev \

Reviewer: You've added all these packages in the single install line above. This vertical list is now duplicated. Please remove it.

# Python

> Gitpod comes with great support for Python builtin. Still, depending on your particular project, you might want to further optimize the experience.

Reviewer: Suggestion to remove unnecessary word:

```suggestion
Gitpod comes with great support for Python builtin. Still, depending on your project, you might want to further optimize the experience.
```

| --------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------: | ------------------------------------------------: |
| [gitpod-io/django-locallibrary-tutorial](https://github.com/gitpod-io/django-locallibrary-tutorial) |        [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#gitpod-io/django-locallibrary-tutorial) |       An example website written in Django by MDN |
| [JesterOrNot/Gitpod-PyQt](https://github.com/JesterOrNot/Gitpod-PyQt)                               |    [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/JesterOrNot/Gitpod-PyQt) |                         A PyQt example for Gitpod |
| [gitpod-io/wxPython-example](https://github.com/gitpod-io/wxPython-example)                         | [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/gitpod-io/wxPython-example) |                     A wxPython example for Gitpod |

> | [techwithtim/Hangman](https://github.com/techwithtim/Hangman)                                       |        [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/techwithtim/Hangman) | A simple hangman game made with Python and PyGame |

Reviewer: I find the center column not ideal for the button. How about moving the `Try it` column to the right, making it `| Name | Description | Try it |`?

Author: Sorry wromg button!

Author: Fixed!

## GUI Applications with wxPython

> To install wxPython to your repository please add the following to your [`.gitpod.Dockerfile`](https://www.gitpod.io/docs/42_config_docker/) if you don't have one please run `gp init` and commit the two generated files.

Reviewer: Suggestion to allow readers to breathe:

```suggestion
To install wxPython to your repository please add the following to your [`.gitpod.Dockerfile`](https://www.gitpod.io/docs/42_config_docker/). If you don't have one, simply run `gp init` and commit the two generated files.
```

```dockerfile
# This will pull the official Gitpod `vnc` image
# which has much of what you need to start
FROM gitpod/workspace-full-vnc

USER gitpod

# Install wxPython dependencies

> RUN sudo apt-get -q update \

Reviewer: I now prefer `&&` at the end (that way you can combine the `&&` and the `\`, making it less likely to accidentally forget one):

```suggestion
RUN sudo apt-get -q update && \
```

>     && sudo DEBIAN_FRONTEND=noninteractive apt-get install -yq \

Reviewer: How about putting all these dependencies into a single line? That could make the Dockerfile appear "less scary", and it's also not really important to call out each of these dependencies:

```suggestion
    sudo DEBIAN_FRONTEND=noninteractive apt-get install -yq freeglut3-dev python3.7-dev libpython3.7-dev libgl1-mesa-dev libglu1-mesa-dev libgstreamer-plugins-base1.0-dev libgtk-3-dev libnotify-dev libsdl2-dev libwebkit2gtk-4.0-dev libxtst-dev libgtk2.0-dev && \
```

    freeglut3-dev \
    python3.7-dev \
    libpython3.7-dev \
    libgl1-mesa-dev \
    libglu1-mesa-dev \
    libgstreamer-plugins-base1.0-dev \
    libgtk-3-dev \
    libnotify-dev \
    libsdl2-dev \
    libwebkit2gtk-4.0-dev \
    libxtst-dev \
    libgtk2.0-dev \

>     && sudo rm -rf /var/lib/apt/lists/*

Reviewer: ```suggestion
    sudo rm -rf /var/lib/apt/lists/*
```

# Install wxPython
RUN pip3 install -U -f https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-18.04/ wxPython
```

> Here is an example `.gitpod.yml`:

Reviewer: Suggestion to explicitly mention there is a link between these examples:

```suggestion
Here is a corresponding `.gitpod.yml` example:
```

```yaml
image:
  file: .gitpod.Dockerfile

# This will expose all necessary ports needed for your VNC image
ports:
  - port: 6080
    onOpen: open-preview
  - port: 5900
    onOpen: ignore
  - port: 35900
    onOpen: ignore

# This will make it so that on workspace start it will run a file called `app.py`
tasks:
  - command: python3 app.py
```
<br>

> We also support other GUI frameworks such as

Reviewer: Such as... what? (This sentence looks incomplete.)

<br>

> Here are some exmaples of other applications in Gitpod

Reviewer: Typo & suggestion:

```suggestion
Here are some other examples of Python GUI applications in Gitpod:
```

> | Name             | Open in Gitpod | Framework |

Reviewer: ```suggestion
| Name             | Framework | Try it |
```

|------------------|----------------|-----------|
| Tic-Tac-Toe-GUI  | [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/JesterOrNot/Tic-Tac-Toe-GUI) | Kivy |
| Pong             | [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/JesterOrNot/Pong) | Kivy |
| Gitpod-PyQt | [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/JesterOrNot/Pong) | PyQt |
<br>

> And here is an example of PyQT running

Reviewer: Sentence looks incomplete.

<br>

## VSCode Extensions

> While most popular Python VSCode extensions are builtin to Gitpod here are a few "Nice to have" extensions that you can use

Reviewer: Phrasing suggestion:

```suggestion
While the most popular Python VSCode extensions are built-in to Gitpod, here are a few "Nice to have" extensions that you can use as well.
```

### ARepl for Python
![](./images/AReplExample.gif)

> It's helpful for constant checking your code and debugging.

Reviewer: ```suggestion
It's helpful for constantly checking your code and debugging.
```

> To add this to your repository add the following to your `.gitpod.yml`

Reviewer: ```suggestion
To install this extension for your repository, add the following to your `.gitpod.yml`:
```

```yaml
vscode:
  extensions:
    - almenon.arepl@1.0.20:Uu7lIOwyLgmNWpTwCl/iqQ==
```
### Python Test Explorer
![](./images/python_Test_In_Gitpod.png)

> Easily test your python extensions.

Reviewer: Please specify what sort of extension you mean here.

# Python

Gitpod comes with great support for Python builtin. Still, depending on your particular project, you might want to further optimize the experience.

## GUI Applications with wxPython
Have a super cool GUI application you are building with wxPython? Want to run it in Gitpod? Please take a look at [JesterOrNot/Gitpod-PythonGUI-Example](https://github.com/JesterOrNot/Gitpod-PythonGUI-Example)

Here is an example `.gitpod.Dockerfile`:
```dockerfile
# This will pull the official Gitpod `vnc` image 
# which has much of what you need to start
FROM gitpod/workspace-full-vnc:latest

USER gitpod

# Install WXPython Dependencies
RUN sudo apt-get -q update \
    && sudo DEBIAN_FRONTEND=noninteractive apt-get install -yq \
    freeglut3-dev \
    python3.7-dev \
    libpython3.7-dev \
    libgl1-mesa-dev \
    libglu1-mesa-dev \
    libgstreamer-plugins-base1.0-dev \
    libgtk-3-dev \
    libnotify-dev \
    libsdl2-dev \
    libwebkit2gtk-4.0-dev \
    libxtst-dev \
    libgtk2.0-dev \

>     && sudo rm -rf /var/lib/apt/lists/*

Reviewer: Note: I still feel like this is too long.

>     && sudo rm -rf /var/lib/apt/lists/*

Author: However many users I assume will just copy paste this dockerfile and it will fail if we don't give them all of it maybe instead of removing it we can do a line by line description of the Dockerfile

>     && sudo rm -rf /var/lib/apt/lists/*

Reviewer: Ok, let's go with the in-line Dockerfile for now.

# Install wxPython
RUN pip3 install -U -f https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-18.04/ wxPython
```

> Here is an example `.gitpod.yml`:

Reviewer: This feels kind of out-of-the-blue: Say I'm a Python developer, and I'm new to Gitpod. I read this guide, see `wxPython` and get excited because that's what I use.

But then I'm pointed to an example repo with a lot of stuff in it (I'm not sure if anything in there can be useful to me, and what I should take from there). And then I'm being thrown some Docker stuff at, and some YAML stuff on top of it, and this is all very overwhelming because I have no idea what to do with all this stuff.

Solution: It would be good to mention what result I may potentially be interested in (e.g. "Installing wxPython"), and then give clear, simple, directive instructions. E.g.

> In order to install `wxPython` for your repository, you can use this command:
> `pip3 install ...`
> 
> If you'd like to automatically install `wxPython` in every Gitpod workspace for your project, you can add a [`.gitpod.Dockerfile`] to your repository, containing the above `pip3` command. Here is a [complete .gitpod.Dockerfile example] you can use that does exactly that.
> 
> Don't forget to also reference this new `.gitpod.Dockerfile` it your [`.gitpod.yml`] configuration file. If you don't already have one, simply create a `.gitpod.yml` file at the root of your repository, and reference your `.gitpod.Dockerfile` like so:
> `image: ...`
> 
> Next you may also want to configure Gitpod's virtual desktop (technically a web server running on port 6080) to be automatically opened in a preview pane. To do this, you can configure your ports like so:
> `ports: ...`
> 
> Finally, you can also make Gitpod run your app automatically on start-up. To do this, simply add these lines to your `.gitpod.yml` file:
> `tasks: ...`
> 
> To conclude, here is the full `.gitpod.yml` we described above:
> `...`
> 
> Once you've committed both `.gitpod.yml` and `.gitpod.Dockerfile` files, you can push them to GitHub. Then, the next time someone opens your project in Gitpod, they will get from the fully-automated setup that you've just configured. Congratulations!

# Python

> 🚧 Please help improve this guide by [editing it](https://gitpod.io/#https://github.com/gitpod-io/website/blob/master/src/docs/python_in_gitpod.md).

Gitpod comes with great support for Python builtin. Still, depending on your particular project, you might want to further optimize the experience.

## GUI Applications with wxPython
Have a super cool GUI application you are building with wxPython? Want to run it in Gitpod? Please take a look at [JesterOrNot/Gitpod-PythonGUI-Example](https://github.com/JesterOrNot/Gitpod-PythonGUI-Example)

Here is an example .gitpod.Dockerfile
```dockerfile
FROM gitpod/workspace-full-vnc:latest

USER root

ARG DEBIAN_FRONTEND=noninteractive

# Install wxPython Dependencies
RUN apt-get -q update \
    && apt-get install -yq \
        freeglut3-dev \
        python3.7-dev \
        libpython3.7-dev \
        libgl1-mesa-dev \
        libglu1-mesa-dev \
        libgstreamer-plugins-base1.0-dev \
        libgtk-3-dev \
        libnotify-dev \
        libsdl2-dev \
        libwebkit2gtk-4.0-dev \
        libxtst-dev \
        libgtk2.0-dev \
    && sudo rm -rf /var/lib/apt/lists/*

# Add wxPython itself

USER gitpod

> RUN pip3 install -U -f https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-18.04/ wxPython

Author: Changing that now

> RUN pip3 install -U -f https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-18.04/ wxPython

Author: Actually this is the only way where it doesn't break the build

> RUN pip3 install -U -f https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-18.04/ wxPython

Reviewer: Works for me then. Thanks for checking!

# Python

Gitpod comes with great support for Python builtin. Still, depending on your particular project, you might want to further optimize the experience.

## Examples

> Before we get started, Here are some examples of already Gitpodified workspace setups!

Reviewer: Typos, suggestion:

```suggestion
Before we get started, here are some examples of already-[gitpodified](https://www.gitpod.io/blog/gitpodify/) repositories!
```

> | Name                                                                                                |                                                                                                              Open In Gitpod Button |                                        Descripton |

Reviewer: Second title is too long:

```suggestion
| Name                                                                                                |                                                                                                              Try it |                                        Descripton |
```
